text/0253-cdk-metadata-v2.md

```diff
@@ -0,0 +1,488 @@
+names currently in monocdk; that set of class names comes to about 23kB (uncompressed).
+
+Trying various methods of compression on it yields the following (keeping in mind
+that the result of binary compression will need to be base64-encoded to transmit via
+a CloudFormation template):
+
+| Description                | Raw size    | Base64ed size |
+|----------------------------|-------------|---------------|
+| Plaintext                  | 23k         | -             |
+| gzipped                    | 3.6k        | 4.8k          |
+| Zopfli                     | 3.4k        | 4.6k          |
+| bzip2                      | 3.4k        | 4.6k          |
+| Affix-grouped (plaintext)  | 7.4k        | -             |
+| Affix-grouped gzipped      | 3.0k        | 4.0k          |
+| Bloom Filter (1% error)    | 720b        | 960b          |
+
+Affix-grouped, gzipped data seems to give the best results that are still
+convenient and generic to work with (see Appendix B for a description of
+prefix grouping).
+
+See Appendix C for an example blob to be added to a template. I'm not super
+happy with this, but this is the most straightfoward option available.
+
+> If we are willing to do more work, we can achieve less space used up by doing
+> more esoteric things. One option would be to use a Bloom Filter, which can use
+> a configurable amount of storage to achieve a configurable hit error ratio.
+> For example, achieving about a 1% false positive rate would take on the order
+> of about ~1k base64-encoded bytes. However, in order to extract a construct list
+> from this Bloom Filter would be a considerable amount of work: we would need
+> a full list of all construct available in every library and at every version,
+> in order to test each construct's presence in the filter. This imposes a lot
+> of complexity on the receiver, which is probably not worth the space savings.
+
+### Analytics payload
+
+The analytics payload itself will be preceded by an encoding scheme identifier,
+followed by a `:` and the transmitted data:
+
+```yaml
+Resources:
+  CDKMetadata:
+    Type: AWS::CDK::Metadata
+    Properties:
+      Analytics: <payload-encoding>:<payload>
+```
+
+The only valid payload encodings at the moment are:
+
+* `deflate64`: base64-encoded, gzipped data.
+* `plain`: plaintext data.
+
+The payload itself is a JSON object with `"constructs"` currently defined as the only
+valid key (but can be extended in the future). Values can be prefixed with
+an encoding scheme as well:
+
+```json
+{
+  "constructs": "[<value-encoding>:]<prefix-encoded construct list>"
+}
+```
+
+The only valid payload value encodings at the moment are:
+
+* `afx`: affix-encoded string list (see Appendix B).
+* `plain`: plaintext data.
+
+Encoding may change the type (the encoded value is a `string` but the decoded
+value is a `string[]`).
+
+> If a value does not start with > /^[a-z0-9]{,10}:/ it is considered to
+> be not encoded (the encoding is considered to be "plain").
```

Comment:
I'd rather consider it is invalid in that case.

If the value does not match expected format, it may well contain data that you did not expect, and may not be safe to drop in logs / storage / etc...